jazkarta/shop/browser/checkout.py

```diff
@@ -24,13 +32,318 @@
+
+        # to debug use http://developer.authorize.net/bin/developer/paramdump
+        # and to prevent this error:
+        # (14) The referrer, relay response or receipt link URL is invalid.
+        # return "http://developer.authorize.net/bin/developer/paramdump"
+        # also useful:
+        # https://support.authorize.net/authkb/index?page=content&id=A663&pmv=print&impressions=false
+
+    @lazy_property
+    def x_cancel_url(self):
+        # for optional cancel button on SIM form. Take the user to the homepage
+        return self.context.absolute_url()
+
+    def handle_submit(self):
+        amount = self.amount
+        response = self.request['x_response_code']
+        if response != '1':
+            # Authorize.Net SIM response codes
+            # 1 = Approved
+            # 2 = Declined
+            # 3 = Error
+            # 4 = Held for Review
+            if 'x_response_reason_text' in self.request.form:
+                self.error = self.request['x_response_reason_text']
+            else:
+                # this scenario unlikely as every response should provide us 
+                # with x_response_reason_text
+                self.error = ('There was an error with your transaction. '
+                              'Your payment has not been processed. '
+                              'Please contact us for assistance.')
+        if self.error:
+            return
+
+        userid = get_current_userid()
+        contact_info = PersistentMapping()
+        if 'x_first_name' in self.request.form:
+            contact_info['first_name'] = self.request.form['x_first_name']
+        if 'x_last_name' in self.request.form:
+            contact_info['last_name'] = self.request.form['x_last_name']
+        if 'x_email' in self.request.form:
+            contact_info['email'] = self.request.form['x_email']
+        if 'x_phone' in self.request.form:
+            contact_info['phone'] = self.request.form['x_phone']
+        if 'x_address' in self.request.form:
+            contact_info['address'] = self.request.form['x_address']
+        if 'x_city' in self.request.form:
+            contact_info['city'] = self.request.form['x_city']
+        if 'x_state' in self.request.form:
+            contact_info['state'] = self.request.form['x_state']
+        if 'x_zip' in self.request.form:
+            contact_info['zip'] = self.request.form['x_zip']
+        if 'x_country' in self.request.form:
+            contact_info['country'] = self.request.form['x_country']
+
+        method = 'Online Payment'
+        charge_result = {'success': True}
+
+        if not self.error:
+            try:
+                self.store_order(method, charge_result, userid, contact_info)
+                self.clear_cart()
+            except ConflictError:
+                raise Exception(
+                    'Failed to store results of payment after multiple '
+                    'retries. Please contact us for assistance.')
```

Comment:
There's not much point in a "Please contact us" message in an exception since regular users will see the generic "there was an error" message, not the specific exception message. I think you want to set self.error here so that the message will be shown in the template.